File: 10. Kruskal.cpp
/*
Task: Kruskal's Algorithm
Find the Minimum Spanning Tree (MST) using the Disjoint Set Union (DSU) data structure.
*/
#include <bits/stdc++.h>
using namespace std;

struct Edge
{
    int u, v, w;
    bool operator<(const Edge& other) const { return w < other.w; }
};
struct DSU
{
    vector<int> parent, rank;
    DSU(int n) : parent(n), rank(n, 0)
    {
        for (int i = 0; i < n; ++i) parent[i] = i;
    }
    int find(int x)
    {
        if (parent[x] != x)
            parent[x] = find(parent[x]);
        return parent[x];
    }
    void unite(int x, int y)
    {
        int xr = find(x), yr = find(y);
        if (xr == yr)
            return;
        if (rank[xr] < rank[yr])
            parent[xr] = yr;
        else if (rank[xr] > rank[yr])
            parent[yr] = xr;
        else
        {
            parent[yr] = xr;
            rank[xr]++;
        }
    }
};
int kruskalMST(int V, vector<Edge>& edges)
{
    sort(edges.begin(), edges.end());
    DSU dsu(V);
    int total = 0;
    for (auto& e : edges)
    {
        if (dsu.find(e.u) != dsu.find(e.v))
        {
            dsu.unite(e.u, e.v);
            total += e.w;
        }
    }
    return total;
}
// Sample Test Case
int main()
{
    int V = 4;
    vector<Edge> edges = {{0, 1, 10}, {0, 2, 6}, {0, 3, 5}, {1, 3, 15}, {2, 3, 4}};
    cout << "Kruskal's MST: " << kruskalMST(V, edges) << endl;
    return 0;
}
/*
Time Complexity: O(ElogE)
Space Complexity: O(V)
*/

File: 10. Kruskal.py
"""
Task: Kruskal's Algorithm
Find the Minimum Spanning Tree (MST) using the Disjoint Set Union (DSU) data structure.
"""
class DSU:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0]*n
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    def unite(self, x, y):
        xr, yr = self.find(x), self.find(y)
        if xr == yr: return
        if self.rank[xr] < self.rank[yr]:
            self.parent[xr] = yr
        elif self.rank[xr] > self.rank[yr]:
            self.parent[yr] = xr
        else:
            self.parent[yr] = xr
            self.rank[xr] += 1
class Edge:
    def __init__(self, u, v, w):
        self.u, self.v, self.w = u, v, w
def kruskal_mst(V, edges):
    edges.sort(key=lambda e: e.w)
    dsu = DSU(V)
    total = 0
    for e in edges:
        if dsu.find(e.u) != dsu.find(e.v):
            dsu.unite(e.u, e.v)
            total += e.w
    return total
# Sample Test Case
if __name__ == "__main__":
    V = 4
    edges = [Edge(0,1,10), Edge(0,2,6), Edge(0,3,5), Edge(1,3,15), Edge(2,3,4)]
    print("Kruskal's MST:", kruskal_mst(V, edges))
"""
Time Complexity: O(ElogE)
Space Complexity: O(V)
"""
